pgrx/src/fcall.rs

```diff
@@ -0,0 +1,461 @@
+
+        if fnoid == pg_sys::InvalidOid {
+            // if that didn't find a function, maybe we've got some defaults in there, so do a lookup
+            // where Postgres will consider that
+            fnoid = pg_sys::LookupFuncName(
+                parts_list.as_mut_ptr(),
+                -1,
+                arg_types.as_ptr(),
+                false, // we want the ERROR here -- could be UNDEFINED_FUNCTION or AMBIGUOUS_FUNCTION
+            );
+        }
+
+        Ok(fnoid)
+    }))
+    .catch_when(PgSqlErrorCode::ERRCODE_INVALID_PARAMETER_VALUE, |_| {
+        Err(FCallError::InvalidIdentifier(fname.to_string()))
+    })
+    .catch_when(PgSqlErrorCode::ERRCODE_AMBIGUOUS_FUNCTION, |_| Err(FCallError::AmbiguousFunction))
+    .catch_when(PgSqlErrorCode::ERRCODE_UNDEFINED_FUNCTION, |_| Err(FCallError::UndefinedFunction))
+    .execute();
+
+    unsafe {
+        // SAFETY:  we palloc'd the `pg_sys::String` elements of `parts_list` above and so it's
+        // safe for us to free them now that they're no longer being used
+        parts_list.drain(..).for_each(|s| {
+            #[cfg(any(feature = "pg11", feature = "pg12", feature = "pg13", feature = "pg14"))]
+            {
+                let s = s.cast::<pg_sys::Value>();
+                pg_sys::pfree((*s).val.str_.cast());
+            }
+
+            #[cfg(any(feature = "pg15", feature = "pg16"))]
+            {
+                let s = s.cast::<pg_sys::String>();
+                pg_sys::pfree((*s).sval.cast());
+            }
+        });
+    }
+
+    result
+}
+
+/// Parses an arbitrary string as if it is a SQL identifier.  If it's not, [`FCallError::InvalidIdentifier`]
+/// is returned
+fn parse_sql_ident(ident: &str) -> Result<Array<&str>> {
+    unsafe {
+        direct_function_call::<Array<&str>>(
+            pg_sys::parse_ident,
+            &[ident.into_datum(), true.into_datum()],
+        )
+        .ok_or_else(|| FCallError::InvalidIdentifier(ident.to_string()))
+    }
+}
+
+/// Materializes a the `DEFAULT` value at the specified argument position `argnum` for the specified
+/// function `pg_proc`.
+///
+/// `argnum` is the overall argument position, not the specific argument from just the set of
+/// arguments with defaults.  This is noted as Postgres internally understands these as different
+/// things.  Given the argument number, [`create_default_value`] determines which argument from the
+/// list of DEFAULTed arguments is being requested.
+///
+/// # Errors
+///
+/// - [`FCallError::NotDefaultArgument`] if the specified `argnum` does not have a `DEFAULT` clause
+/// - [`FCallError::DefaultNotConstantExpression`] if the `DEFAULT` clause is one we cannot evaluate
+fn create_default_value(pg_proc: &PgProc, argnum: usize) -> Result<Option<pg_sys::Datum>> {
+    let non_default_args_cnt = pg_proc.pronargs() - pg_proc.pronargdefaults();
+    if argnum < non_default_args_cnt {
+        return Err(FCallError::NotDefaultArgument(argnum));
+    }
```

Comment:
Is it the case that default arguments must only come last? What about functions with mixed or named notation?

Reply:
yes, default arguments must come last.

None of this supported named notation -- I should have listed that in the "things we don't support" up above.